### ISSUE TITLE
ActiveStorage throws NoMethodError when using MiniMagick's "format"

### ISSUE THREAD

**Issue by brenogazzola** [rails/rails#32063]
According to Active Storage's guide, it supports all of MiniMagick's transformations. MiniMagicks' first usage example is one combining resize and format. However, format is the only transformation I couldn't get to work so far. Rotate, crop, resize all worked fine.

### Steps to reproduce
1 - Attach a png image;
2 - Request a variant `.variant(format:"JPEG")`;

### Expected behavior
A JPEG variant is generated.

### Actual behavior
Rails throws a NoMethodError.

### System configuration
**Rails version**:
5.2.0.rc1
**Ruby version**:
2.4.3 and 2.5.0
**MiniMagick version**:
4.8.0

### Relevant Code ###
```
class Photo < ApplicationRecord
  belongs_to :event
  has_one_attached :file
end

<%= image_tag @event.photo.file.variant(format:"JPEG") %>
<%= image_tag @event.photo.file.variant(format:"JPG") %>
```
### Stack Trace
```
NoMethodError (you must call #format on a MiniMagick::Image directly):
  
mini_magick (4.8.0) lib/mini_magick/tool/mogrify_restricted.rb:10:in `format'
activestorage (5.2.0.rc1) app/models/active_storage/variation.rb:72:in `public_send'
activestorage (5.2.0.rc1) app/models/active_storage/variation.rb:72:in `pass_transform_argument'
activestorage (5.2.0.rc1) app/models/active_storage/variation.rb:57:in `block (3 levels) in transform'
mini_magick (4.8.0) lib/mini_magick/image.rb:565:in `block in mogrify'
mini_magick (4.8.0) lib/mini_magick/tool.rb:39:in `new'
mini_magick (4.8.0) lib/mini_magick/image.rb:564:in `mogrify'
activestorage (5.2.0.rc1) app/models/active_storage/variation.rb:51:in `block (2 levels) in transform'
activestorage (5.2.0.rc1) app/models/active_storage/variation.rb:50:in `each'
activestorage (5.2.0.rc1) app/models/active_storage/variation.rb:50:in `block in transform'
activesupport (5.2.0.rc1) lib/active_support/notifications.rb:170:in `instrument'
activestorage (5.2.0.rc1) app/models/active_storage/variation.rb:49:in `transform'
activestorage (5.2.0.rc1) app/models/active_storage/variant.rb:122:in `transform'
activestorage (5.2.0.rc1) app/models/active_storage/variant.rb:86:in `block in process'
activestorage (5.2.0.rc1) app/models/active_storage/variant.rb:110:in `open_image'
activestorage (5.2.0.rc1) app/models/active_storage/variant.rb:85:in `process'
activestorage (5.2.0.rc1) app/models/active_storage/variant.rb:53:in `processed'
activestorage (5.2.0.rc1) app/controllers/active_storage/variants_controller.rb:12:in `show'
actionpack (5.2.0.rc1) lib/action_controller/metal/basic_implicit_render.rb:6:in `send_action'
actionpack (5.2.0.rc1) lib/abstract_controller/base.rb:194:in `process_action'
actionpack (5.2.0.rc1) lib/action_controller/metal/rendering.rb:30:in `process_action'
actionpack (5.2.0.rc1) lib/abstract_controller/callbacks.rb:42:in `block in process_action'
activesupport (5.2.0.rc1) lib/active_support/callbacks.rb:132:in `run_callbacks'
actionpack (5.2.0.rc1) lib/abstract_controller/callbacks.rb:41:in `process_action'
actionpack (5.2.0.rc1) lib/action_controller/metal/rescue.rb:22:in `process_action'
actionpack (5.2.0.rc1) lib/action_controller/metal/instrumentation.rb:34:in `block in process_action'
activesupport (5.2.0.rc1) lib/active_support/notifications.rb:168:in `block in instrument'
activesupport (5.2.0.rc1) lib/active_support/notifications/instrumenter.rb:23:in `instrument'
activesupport (5.2.0.rc1) lib/active_support/notifications.rb:168:in `instrument'
actionpack (5.2.0.rc1) lib/action_controller/metal/instrumentation.rb:32:in `process_action'
actionpack (5.2.0.rc1) lib/action_controller/metal/params_wrapper.rb:256:in `process_action'
activerecord (5.2.0.rc1) lib/active_record/railties/controller_runtime.rb:24:in `process_action'
actionpack (5.2.0.rc1) lib/abstract_controller/base.rb:134:in `process'
actionview (5.2.0.rc1) lib/action_view/rendering.rb:32:in `process'
actionpack (5.2.0.rc1) lib/action_controller/metal.rb:191:in `dispatch'
actionpack (5.2.0.rc1) lib/action_controller/metal.rb:252:in `dispatch'
actionpack (5.2.0.rc1) lib/action_dispatch/routing/route_set.rb:52:in `dispatch'
actionpack (5.2.0.rc1) lib/action_dispatch/routing/route_set.rb:34:in `serve'
actionpack (5.2.0.rc1) lib/action_dispatch/journey/router.rb:52:in `block in serve'
actionpack (5.2.0.rc1) lib/action_dispatch/journey/router.rb:35:in `each'
actionpack (5.2.0.rc1) lib/action_dispatch/journey/router.rb:35:in `serve'
actionpack (5.2.0.rc1) lib/action_dispatch/routing/route_set.rb:840:in `call'
omniauth (1.8.1) lib/omniauth/strategy.rb:190:in `call!'
omniauth (1.8.1) lib/omniauth/strategy.rb:168:in `call'
omniauth (1.8.1) lib/omniauth/strategy.rb:190:in `call!'
omniauth (1.8.1) lib/omniauth/strategy.rb:168:in `call'
omniauth (1.8.1) lib/omniauth/builder.rb:63:in `call'
rack (2.0.4) lib/rack/tempfile_reaper.rb:15:in `call'
rack (2.0.4) lib/rack/etag.rb:25:in `call'
rack (2.0.4) lib/rack/conditional_get.rb:25:in `call'
rack (2.0.4) lib/rack/head.rb:12:in `call'
actionpack (5.2.0.rc1) lib/action_dispatch/http/content_security_policy.rb:18:in `call'
rack (2.0.4) lib/rack/session/abstract/id.rb:232:in `context'
rack (2.0.4) lib/rack/session/abstract/id.rb:226:in `call'
actionpack (5.2.0.rc1) lib/action_dispatch/middleware/cookies.rb:663:in `call'
activerecord (5.2.0.rc1) lib/active_record/migration.rb:558:in `call'
actionpack (5.2.0.rc1) lib/action_dispatch/middleware/callbacks.rb:28:in `block in call'
activesupport (5.2.0.rc1) lib/active_support/callbacks.rb:98:in `run_callbacks'
actionpack (5.2.0.rc1) lib/action_dispatch/middleware/callbacks.rb:26:in `call'
actionpack (5.2.0.rc1) lib/action_dispatch/middleware/executor.rb:14:in `call'
actionpack (5.2.0.rc1) lib/action_dispatch/middleware/debug_exceptions.rb:61:in `call'
web-console (3.5.1) lib/web_console/middleware.rb:135:in `call_app'
web-console (3.5.1) lib/web_console/middleware.rb:28:in `block in call'
web-console (3.5.1) lib/web_console/middleware.rb:18:in `catch'
web-console (3.5.1) lib/web_console/middleware.rb:18:in `call'
actionpack (5.2.0.rc1) lib/action_dispatch/middleware/show_exceptions.rb:33:in `call'
railties (5.2.0.rc1) lib/rails/rack/logger.rb:38:in `call_app'
railties (5.2.0.rc1) lib/rails/rack/logger.rb:26:in `block in call'
activesupport (5.2.0.rc1) lib/active_support/tagged_logging.rb:71:in `block in tagged'
activesupport (5.2.0.rc1) lib/active_support/tagged_logging.rb:28:in `tagged'
activesupport (5.2.0.rc1) lib/active_support/tagged_logging.rb:71:in `tagged'
railties (5.2.0.rc1) lib/rails/rack/logger.rb:26:in `call'
sprockets-rails (3.2.1) lib/sprockets/rails/quiet_assets.rb:13:in `call'
actionpack (5.2.0.rc1) lib/action_dispatch/middleware/remote_ip.rb:81:in `call'
actionpack (5.2.0.rc1) lib/action_dispatch/middleware/request_id.rb:27:in `call'
rack (2.0.4) lib/rack/method_override.rb:22:in `call'
rack (2.0.4) lib/rack/runtime.rb:22:in `call'
activesupport (5.2.0.rc1) lib/active_support/cache/strategy/local_cache_middleware.rb:29:in `call'
actionpack (5.2.0.rc1) lib/action_dispatch/middleware/executor.rb:14:in `call'
actionpack (5.2.0.rc1) lib/action_dispatch/middleware/static.rb:127:in `call'
rack (2.0.4) lib/rack/sendfile.rb:111:in `call'
railties (5.2.0.rc1) lib/rails/engine.rb:524:in `call'
puma (3.11.2) lib/puma/configuration.rb:225:in `call'
puma (3.11.2) lib/puma/server.rb:624:in `handle_request'
puma (3.11.2) lib/puma/server.rb:438:in `process_client'
puma (3.11.2) lib/puma/server.rb:302:in `block in run'
puma (3.11.2) lib/puma/thread_pool.rb:120:in `block in spawn_thread'
```

**Comment by brenogazzola** [comment]
I took a look at the Variant model, and noticed that it has a `transform` method that probably would fit my need. Unfortunately it is private and it seems to only be used to take png images of files that are not images. 

For now I'm solving my problem by going through MiniMagick, but it would be nice if ActiveStorage could it for us. 

This is the code I'm using in the job that generates the variants.
````
if photo.file.content_type != "image/jpeg"
  image = MiniMagick::Image.open(url_for(photo.file))
  image.format "jpeg"
  photo.file.attach(io: File.open(image.path), filename: photo.file.filename.to_s, content_type: "image/jpg")
end
````

**Comment by georgeclaghorn** [comment]
Thanks for the report, but variants are unlikely to support converting images to a specified format without a compelling reason. They’re intended for visibly transforming images.

**Comment by johnlpollard** [comment]
@brenogazzola Where did you put that code to force jpeg? I have a requirement that the images must be jpeg due to the consumers of the images.

**Comment by brenogazzola** [comment]
@johnlpollard I added it to a job that I enqueue after the record that contains the image is committed, so that the variants are generated before they are needed. 

Here's the full code (I changed the one above).

````
class ApplicationJob < ActiveJob::Base
  include Rails.application.routes.url_helpers

  class << self
    def default_url_options
      Rails.application.config.action_mailer.default_url_options
    end
  end
end


class GeneratePhotoVariantsJob < ApplicationJob
  queue_as :default

  def perform(photo)
    if photo.file.content_type != "image/jpeg"
      image = MiniMagick::Image.open(url_for(photo.file))
      image.format "jpeg"
      photo.file.attach(io: File.open(image.path), filename: File.basename(image.path), content_type: "image/jpeg")
    end

    GeneratePhotoVariantsMobileJob.perform_later(photo)
    GeneratePhotoVariantsDesktopJob.perform_later(photo)
    GeneratePhotoVariantsPreviewJob.perform_later(photo)
    GeneratePhotoVariantsOpengraphJob.perform_later(photo)
  end
end
````

**Comment by kbroderick** [comment]
@georgeclaghorn what about creating web previews of non-web-safe formats (e.g. PSD or EPS)? For an app that works with print-ready files, being able to create a web-viewable preview is important, but you don't want to replace a print-quality (and presumably CMYK) PSD with a JPEG, which is what the workaround posted above would do.

**Comment by TheNeikos** [comment]
@georgeclaghorn can you elaborate on why this would be a bad feature to have? 

Generating a 'low quality' preview (JPG) of a high quality PNG sounds to me like just the thing you described, which is an alternate representation ('visibly transforming') of the attached data.

**Comment by MatteoJoliveau** [comment]
Still no news about this?
We have the need to optimize images uploaded from the application by creating scaled and thumbnail versions in JPEG 2000, having it directly set in the variant configuration would be ideal

**Comment by Petercopter** [comment]
Another use case would be for converting animated gifs to first frame previews. I've got a bunch of animated gifs that were optimized with gifsicle, and Imagemagick is not able to resize them correctly. What I was looking to do is save the first frame as a PNG.

**Comment by bernb** [comment]
I would also love to have variants support format change.
My use case is that TIFFs are uploaded and I want to show a preview of those (most browser do not render TIFFs).

**Comment by brenogazzola** [comment]
I don't think they will make this change, unfortunately. 

For those looking, this is my job to convert the images currently looks, updated to image processing, handling gifs, and really small images.

```
# frozen_string_literal: true

class GeneratePhotoVariantsJob < ApplicationJob
  queue_as :default

  def perform(photo)
    if photo.file.content_type != "image/jpeg" && photo.file.content_type != "image/jpg" && photo.file.byte_size > 25000
        file = download(photo)
        image = ImageProcessing::MiniMagick.source(file).convert("jpg").loader(page: 1).call
        photo.file.attach(io: File.open(image.path), filename: File.basename(image.path), content_type: "image/jpeg")
      end
  end

  private
    def download(photo)
      io = open(url_for(photo.file))

      if io.is_a?(StringIO)
        downloaded = Tempfile.new
        File.write(downloaded.path, io.string)
      else
        downloaded = io
      end

      downloaded
    end
end
```

**Comment by collimarco** [comment]
I would love to see this feature implemented in ActiveStorage `variant`... It really makes sense to have this format "variant" implemented  there. 

**Converting from PNG to JPEG would allow us to properly compress the images** and drastically reduce cost of bandwidth (we could **reduce the cost up to 80%, which is impressive**). 

Note: With Paperclip it was possible to convert the format and it was also possible to use optimizers like pngquant. Now ActiveStorage does not offer these basic features.

**Comment by brenogazzola** [comment]
This seems to have been implemented in #40226. A new test demonstrates active storage converting a JPEG to PNG.

```
test "PNG variation of JPEG blob" do
  blob = create_file_blob(filename: "racecar.jpg")
  variant = blob.variant(format: :png).processed
  assert_equal "racecar.png", variant.filename.to_s
  assert_equal "image/png", variant.content_type
  assert_equal "PNG", read_image(variant).type
end
```

**Comment by brenogazzola** [comment]
Meanwhile, anyone using ImageProcessing can cheat using the `convert` keyword instead of `format`.

```
def image_for_web(attachment, resize: "1200x")
  attachment.variant { 
    resize: resize
    sampling_factor: "4:2:0",
    quality: 85,
    strip: true,
    interlace: "JPEG",
    colorspace: "sRGB",
    convert: ("jpg" unless file.content_type.in?(["image/jpeg", "image/jpg"]))
  }.compact
end
```
